pkg/image/importer/importer.go

```diff
@@ -0,0 +1,739 @@
+				if digest.Err != nil {
+					setImageImportStatus(isi, index, digest.Err)
+					continue
+				}
+				image := &isi.Status.Images[index]
+				copied := *digest.Image
+				ref := repo.Ref
+				ref.Tag, ref.ID = "", copied.Name
+				copied.DockerImageReference = ref.MostSpecific().Exact()
+				image.Image = &copied
+				image.Status.Status = unversioned.StatusSuccess
+			}
+		}
+	}
+}
+
+func importFromRepository(ctx *AuthContext, isi *api.ImageStreamImport, maximumTags int, cache map[location]*api.Image, limiter util.RateLimiter) {
+	if isi.Spec.Repository == nil {
+		return
+	}
+	isi.Status.Repository = &api.RepositoryImportStatus{}
+	status := isi.Status.Repository
+
+	spec := isi.Spec.Repository
+	from := spec.From
+	if from.Kind != "DockerImage" {
+		return
+	}
+	ref, err := api.ParseDockerImageReference(from.Name)
+	if err != nil {
+		status.Status = invalidStatus("", fielderrors.NewFieldInvalid("from.name", from.Name, fmt.Sprintf("invalid name: %v", err)))
+		return
+	}
+	defaultRef := ref.DockerClientDefaults()
+	repoName := defaultRef.RepositoryName()
+	registryURL := defaultRef.RegistryURL()
+
+	l := location{*registryURL, repoName, ""}
+	repo := &importRepository{
+		Ref:         ref,
+		Registry:    &l.url,
+		Name:        l.name,
+		Insecure:    spec.ImportPolicy.Insecure,
+		MaximumTags: maximumTags,
+	}
+	importRepositoryFromDocker(ctx, repo, limiter)
+
+	if repo.Err != nil {
+		status.Status = imageImportStatus(repo.Err, "", "repository")
+		return
+	}
+
+	additional := []string{}
+	tagLocation := l
+	for _, s := range repo.AdditionalTags {
+		tagLocation.value = s
+		if image, ok := cache[tagLocation]; ok {
+			repo.Tags = append(repo.Tags, importTag{
+				Name:  s,
+				Image: image,
+			})
+		} else {
+			additional = append(additional, s)
+		}
+	}
+	status.AdditionalTags = additional
+
+	failures := false
+	status.Status.Status = unversioned.StatusSuccess
+	status.Images = make([]api.ImageImportStatus, len(repo.Tags))
+	for i, tag := range repo.Tags {
+		if tag.Err != nil {
+			failures = true
+			status.Images[i].Status = imageImportStatus(tag.Err, "", "repository")
+			continue
+		}
+		status.Images[i].Status.Status = unversioned.StatusSuccess
+
+		copied := *tag.Image
+		ref.Tag, ref.ID = tag.Name, copied.Name
+		copied.DockerImageReference = ref.MostSpecific().Exact()
+		status.Images[i].Tag = tag.Name
+		status.Images[i].Image = &copied
+	}
+	if failures {
+		status.Status.Status = unversioned.StatusFailure
+		status.Status.Reason = unversioned.StatusReason("ImportFailed")
+		status.Status.Message = "one or more images from this repository failed to import"
+	}
+}
+
+func applyErrorToRepository(repository *importRepository, err error) {
+	repository.Err = err
+	for i := range repository.Tags {
+		repository.Tags[i].Err = err
+	}
+	for i := range repository.Digests {
+		repository.Digests[i].Err = err
+	}
+}
+
+func importRepositoryFromDocker(ctx *AuthContext, repository *importRepository, limiter util.RateLimiter) {
```

Comment:
describe how errors are returned so most people don't have read through impl and down the call chain.